### PR TITLE
WIP: handle delta locus object 

### PR DESCRIPTION
**Pull request by arun3528**
# Pull Request Template
Handles delta locus and adds other locus object
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Test Coverage

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* SDK Version
* Node/Browser Version
* NPM Version

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  updateLocusUrl(url) {
+    if (url && this.url !== url) {
+      this.url = url;
+      console.info('LocusInfo:url got updated');
+    }
+  }
+
+  updateaclUrl(aclUrl) {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js]: typo

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+  addedOrRemovedParticipant() {
+    return {
+      addedParticipants: [],
+      removedParticipants: [],
+      updatedParticipants: []
+    };
+  }
+
+  updateParticipants(participants, deltaParticpantFlag) {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js]: can we split this function up?

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import LocusUtil from '../locus-events/utils';
+import {DELTAEVENT} from '../constants';
+
+const Parser = {};
+
+// https://sqbu-github.cisco.com/WebExSquared/cloud-apps/wiki/Locus-Delta-Events
+Parser.updateWithDeltaLocus = (deltaLocus, meeting) => {
+  // /* istanbul ignore else */
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: can we remove the dead code

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  }
+  console.log('RESULT :', res);
+  switch (res) {
+    case DELTAEVENT.LT:
+    case DELTAEVENT.EQ:
+      meeting.locusDesync = false;
+      meeting.needToGetFullLocus = false;
+      break;
+    case DELTAEVENT.CF:
+    // uuuujhkhkjhkjhkhkj
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: remove

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      LocusUtil.generateSyncDebugFlag(meeting, deltaLocus);
+      if (meeting.locusDesync) {
+        // meeting.needToGetFullLocus = true;
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: remove?

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  // In-coming locus is greater than working copy, need to compare with base if there
+  // is base sequence except the first delta locus event
+  if (res === DELTAEVENT.GT && deltaLocus.baseSequence && meeting.locus.syncUrl) {
+    resBase = LocusUtil.compareLocus(deltaLocus.baseSequence, meeting.locus.sequence);
+    if (resBase === DELTAEVENT.LT || resBase === DELTAEVENT.CF) {
+      res = DELTAEVENT.CF;
+    }
+  }
+  // Special case for OBTP with meeting size equal or over 5. Locus might change this later
+  // This is a temporary fix. I will find a better solution later.
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: can we get some more notes on this?

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      }
+      else {
+        meeting.locusDesync = true;
+        Parser.syncMeeting(meeting, deltaLocus);
+      }
+      break;
+    case DELTAEVENT.GT:
+      console.info('MeetingAmpstate#updateWithDeltaLocus: MeetingConstants.deltaLocusResult.GT');
+      meeting.locusInfo.onDeltaLocus(deltaLocus);
+      // Parser.updateMeeting(deltaLocus, meeting);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: remove?

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+      // Turn off desync in case this delta locus comes from a sync request
+      meeting.locusDesync = false;
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: are we using this?

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  return res;
+};
+
+
+Parser.syncMeeting = (meeting, deltaLocus) => {
+  if (meeting
+        && meeting.locus
+        && meeting.locus.syncUrl
+        && meeting.locus.fullState
+        && meeting.locus.fullState.state !== 'inactive') {
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: constantize inactive

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js [hunk 10]
```diff
@@ -1,20 +1,35 @@
...
-   * @param {Object} options
-   */
-  constructor(attrs, options) {
-    super({}, options);
+
+  onDeltaLocus(locus, meeting) {
+    console.log('Its a delta event');
+    let res;
+    if (!meeting) {
+    // check for the existing meeting object
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js]: mark with TODO:

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js [hunk 11]
```diff
@@ -1,20 +1,35 @@
...
+      console.log('meeting object not present');
+      console.log('Locus Url present ', locus.locusUrl);
+      //  res = meeting.updateWithDeltaLocus(locus, true);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js]: remove?

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js [hunk 12]
```diff
@@ -1,20 +1,35 @@
...
+    else {
+      res = Parser.updateWithDeltaLocus(locus, meeting);
+    }
+
+    /* istanbul ignore else */
+    if (res === DELTAEVENT.GT
+      || res === DELTAEVENT.CF) {
+      // TODO:  there is a change emit and event
+    }
+    // console.log(`LOCUS DELTA EVENT ${data}`);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js]: uncomment?

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js [hunk 13]
```diff
@@ -27,34 +42,45 @@ export default class LocusParser extends StatelessSparkPlugin {
...
       case LOCUSEVENT.PARTICIPANT_ROLES_UPDATED:
-        console.log('participantUpdated', eventType);
-        //   this.participantUpdated(data);
+        console.log('participant-ROLES_UPDATED', eventType);
         break;
 
       case LOCUSEVENT.PARTICIPANT_DECLINED:
-        console.log('participantDeclined', eventType);
+        console.log('participant_DECLINED', eventType);
         //   this.participantDeclined(data);
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/parser.js]: remove

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    if (!sequence) {
+      return 'undef';
+    }
+    return (ParseUtils.isEmptySequence(sequence)) ? 'empty' : 'def';
+  };
+  flags[0] = setFlag(baseSequence);
+  flags[1] = setFlag(workingSequence);
+  flags[2] = setFlag(targetSequence);
+
+  flags[3] = (baseSequence && workingSequence) ? ParseUtils.compareLocus(baseSequence, workingSequence).toLowerCase() : 'na';
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js]: constantize na

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+  const {baseSequence} = deltaLocus;
+  const workingSequence = meeting.locus.sequence;
+  const targetSequence = deltaLocus.sequence;
+  const flags = ['', '', '', '', '', ''];
+
+  const setFlag = (sequence) => {
+    if (!sequence) {
+      return 'undef';
+    }
+    return (ParseUtils.isEmptySequence(sequence)) ? 'empty' : 'def';
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js]: constantize

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    if ((entry < s2.rangeStart || entry > s2.rangeEnd) && (s2.entries.indexOf(entry) === -1)) {
+      res.push(entry);
+    }
+  });
+  return res;
+};
+
+
+ParseUtils.isEmptySequence = (sequence) => {
+  if (Object.prototype.hasOwnProperty.call(sequence, 'rangeStart') && sequence.rangeStart === 0
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js]: constantize

File: packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    && Object.prototype.hasOwnProperty.call(sequence, 'rangeEnd') && sequence.rangeEnd === 0
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/locus-events/utils.js]: constantize

File: packages/node_modules/@webex/plugin-meetings/src/meeting/request.js [hunk 18]
```diff
@@ -78,6 +78,57 @@ export default class MeetingRequest extends StatelessSparkPlugin {
...
+   * @returns {Promise}
+   */
+
+  getFullLocus(options) {
+    let {locusUrl} = options;
+    const {desync} = options;
+
+    if (locusUrl) {
+      if (desync) {
+        locusUrl += `?sync_debug=${desync}`;
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/request.js]: constantize sync_debug not ? and =

File: packages/node_modules/@webex/plugin-meetings/src/meeting/request.js [hunk 19]
```diff
@@ -78,6 +78,57 @@ export default class MeetingRequest extends StatelessSparkPlugin {
...
+      }
+      return this.request({
+        method: 'GET',
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/request.js]: GET is a constant

File: packages/node_modules/@webex/plugin-meetings/src/roap/util.js [hunk 20]
```diff
@@ -34,14 +34,43 @@ RoapUtil.ensureMeeting = (meeting, type) => {
...
 };
 
-RoapUtil.updatePeerConnection = (meeting, session) =>
-  PeerConnectionManager.updatePeerConnection({
+RoapUtil.compareLastRemoteOffer = (currentOffer, previousOffer) => {
+  let index1, index2;
+  let difference = false;
+
+  if (previousOffer.length > 0 && previousOffer.length === currentOffer.length) {
+    console.log('previous offer', previousOffer[0].slice(0, 20));
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/roap/util.js]: can you mark why 0 - 20? and maybe constantize for more descriptive

File: packages/node_modules/@webex/plugin-meetings/src/roap/util.js [hunk 21]
```diff
@@ -34,14 +34,43 @@ RoapUtil.ensureMeeting = (meeting, type) => {
...
+    console.log('current offer', currentOffer[0].slice(0, 20));
+    previousOffer.forEach((sdp, index) => {
+      index1 = sdp.indexOf('IN');
```
Comment (chrisadubois) [inline packages/node_modules/@webex/plugin-meetings/src/roap/util.js]: constantize

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js (hunk 1 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js]: there is something called acl URl i was also not sure

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js (hunk 2 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-Info/index.js]: We can do that in the next PR it need more changes to it

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js (hunk 6 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: benrui added it it. we will change it when we implement obtp

File: packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js (hunk 8 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-Info/parser.js]: ya

**Pull request closed without merging**